### ISSUE TITLE
Enum name validation ArgumentOutOfRangeException

### ISSUE THREAD

**Issue by jhoward321** [graphql-dotnet/graphql-dotnet#1949]
## Description

I have an enum (that I don't control) that I need to map to names from an EnumMember attribute on the enum. Which looks like something this.

```
public enum BasicType
  {
    [EnumMember(Value = "file")] File,
    [EnumMember(Value = "folder")] Folder,
    [Obsolete("Kit is no longer valid"), EnumMember(Value = "kit")] Kit,
    [Obsolete("Template is no longer valid"), EnumMember(Value = "template")] Template,
    [EnumMember(Value = "url")] Url,
    [EnumMember(Value = "article")] Article,
    [EnumMember(Value = "email-template-component")] EmailTemplateComponent,
    [EnumMember(Value = "email-template")] EmailTemplate,
  }
```

In order to do this I created a custom enumeration graph type which looks like this. This works for what I want, except when adding the enum value names containing dashes, it fails validation in NameValidator.ValidateName(value.Name, "enum");

The regex supplied does not allow dashes. In order to be compatible with some other apis I need to be able to use the dashes in the name. Is this something that's possible?

```
public class BasicTypeEnum : EnumerationGraphType
    {
        public BasicTypeEnum()
        {
            Name = nameof(BasicType);
            Description = typeof(BasicType).Description();
            DeprecationReason = typeof(BasicType).ObsoleteMessage();
            var enumValues = Enum.GetValues(typeof(BasicType));
            foreach (var value in enumValues)
            {
                var name = Enum.GetName(typeof(BasicType), value);
                var correctedName = typeof(BasicType).GetMember(name).First().GetCustomAttribute<EnumMemberAttribute>()
                    .Value ?? name.ToLowerInvariant();
                var obsoleteReason = typeof(BasicType).GetMember(name).First().GetCustomAttribute<ObsoleteAttribute>()
                    ?.Message;
                AddValue(correctedName, null, value, obsoleteReason);
            }
        }
    }
```

### Steps to reproduce

Create an enumeration graph type for an enum with EnumMember attributes with dashes in name. Code above should reproduce it.

### Expected result

I should be able to have a graphql enum name containing dashes.

### Actual result

ArgumentOutOfRangeException thrown in NameValidator.ValidateName

### Environment

Running graphql v3.0.0.2026

**Comment by sungam3r** [comment]
I was once interested in this question myself. I no longer remember what was the reason. The answer is very simple - http://spec.graphql.org/June2018/#sec-Names:
![изображение](https://user-images.githubusercontent.com/21261007/97235360-c1274480-17f3-11eb-933d-6b99c62db4fd.png)

So I think that the answer in _no, you can not do it by means of this package_.

**Comment by sungam3r** [comment]
`EnumMember` is used for different purpose. This is an implementation detail. Do not use values from this attribute into your public GraphQL API.

**Comment by Shane32** [comment]
Once we have a fix for #1840, then an `INameConverter` could be written to remove the dashes from the name.

**Comment by jhoward321** [comment]
Yeah I've seen that from other issues on here. I actually found a way to bypass the validation check by manually adding my names to Values but this had the issue of breaking graphql playground, presumably because of the specified naming conventions. Under normal circumstances I wouldn't care about having to follow the conventions, but since normal json serialization takes the dashes from EnumMember, I am trying to maintain close compatibility with an existing api this is replacing. Now the downstream consumers (who have the same enum) have to do a tostring instead of just passing the value like before which is unfortunate. Regardless I have a solution, but I was hoping to have a clean transition for my consumers.

**Comment by jhoward321** [comment]
The INameConverter solution sounds very interesting and something I would take advantage of if #1840 is fixed

**Comment by sungam3r** [comment]
@jhoward321 Validation happens here:
```c#
   public void AddValue(EnumValueDefinition value)
        {
            if (value == null)
                throw new ArgumentNullException(nameof(value));

            NameValidator.ValidateName(value.Name, "enum");
            Values.Add(value);
        }
```
but `Values` is public:
```c#
 public EnumValues Values { get; }
```
so I think that you already can just write your own `BasicTypeEnum` as you mentioned above with just one replacement:
`AddValue(correctedName, null, value, obsoleteReason);`
->
` Values.Add(new EnumValueDefinition
            {
                Name = correctedName,
                Description = null,
                Value = value,
                DeprecationReason = obsoleteReason
            });`

**Comment by jhoward321** [comment]
Yeah I actually found the same workaround, but it seems to break graphql playground so unfortunately it's a non starter for me

**Comment by sungam3r** [comment]
We can do nothing with external validation in tooling. Try different tool - GraphiQL, Altair. It is interesting for me too.

**Comment by MahdiElahi** [comment]
Hi

i want set name for fields but cant set out of this pattern /^[_a-zA-Z][_a-zA-Z0-9]*$

Field(x=>x.Description).Name("حساب")

my field name is : حساب

how can i change this pattern ??

**Comment by Shane32** [comment]
Fixed in #1961 and #1964 - there is now a way to bypass name validation
